res/controllers/midi-components-0.1.js

```diff
@@ -0,0 +1,1135 @@
+                        engine.setValue(this.group, "cue_gotoandplay", 1);
+                    }
+                }
+            };
+        }
+        shift() {
+            this.input = function(channel, control, value, status, _group) {
+                if (this.isPress(channel, control, value, status)) {
+                    if (engine.getValue(this.group, "play") === 1) {
+                        engine.setValue(this.group, "play", 0);
+                    } else {
+                        engine.setValue(this.group, "eject", 1);
+                    }
+                } else {
+                    if (engine.getValue(this.group, "play") === 0) {
+                        engine.setValue(this.group, "eject", 0);
+                    }
+                }
+            };
+        }
+        output(_value, _group, _control) {
+            if (engine.getValue(this.group, "track_loaded") === 1) {
+                if (this.loaded === undefined) {
+                    this.send(this.on);
+                } else {
+                    if (engine.getValue(this.group, "play") === 1) {
+                        if (this.looping !== undefined &&
+                            engine.getValue(this.group, "repeat") === 1) {
+                            this.send(this.looping);
+                        } else {
+                            this.send(this.playing);
+                        }
+                    } else {
+                        this.send(this.loaded);
+                    }
+                }
+            } else {
+                if (this.empty === undefined) {
+                    this.send(this.off);
+                } else {
+                    this.send(this.empty);
+                }
+            }
+        }
+        connect() {
+            this.connections[0] = engine.makeConnection(this.group, "track_loaded", this.output.bind(this));
+            if (this.playing !== undefined) {
+                this.connections[1] = engine.makeConnection(this.group, "play", this.output.bind(this));
+            }
+            if (this.looping !== undefined) {
+                this.connections[2] = engine.makeConnection(this.group, "repeat", this.output.bind(this));
+            }
+        }
+    };
+
+    const EffectAssignmentButton = class extends Button {
+        constructor(options) {
+            options.key = "group_" + options.group + "_enable";
+            options.group = "[EffectRack1_EffectUnit" + options.effectUnit + "]";
+            options.type = Button.types.toggle;
+            super(options);
+        }
+    };
+
+    const Pot = class Pot extends Component {
+        constructor(options) {
+            super(options);
+            this.firstValueReceived = false;
+        }
+        input(channel, control, value, _status, _group) {
+            if (this.MSB !== undefined) {
+                value = (this.MSB << 7) + value;
+            }
+            let newValue = this.inValueScale(value);
+            if (this.invert) {
```

Comment:
we probably need to do value clamping before doing this inversion to protect against bad input